README
NAME
    Nagios::Plugin::RRD - Create RRD threshold Nagios plugins

SYNOPSIS
     use strict;
     use Nagios::Plugin::RRD qw();
     
 my $rrd = Nagios::Plugin::RRD->new(
             rrd => "/var/spool/rrd/host1.acme.com/mailq.rrd"
         );
     
 $rrd->add_rule(
             ds => "",
             threshold => "",
             result => ""
         );
     
 my ($rtn,$msg) = $rrd->result;

DESCRIPTION
METHODS
  new
     my $rrd = Nagios::Plugin::RRD->new(
             rrd => "/path/filename.rrd",
             die_on_unknown => 1,
         );

  last
     my $last_updated = $rrd->last;

  sources
     my @sources = $rrd->sources;

  add_rule
     my @sources = $rrd->sources;
     for my $ds (@sources) {
         $rrd->add_rule(
                 ds => $ds,
                 rule => "",
                 result => "CRITICAL",
             );
     }

  result
     my ($rtn,$msg) = $rrd->result;
     print "$msg\n";
     exit $rtn;

SEE ALSO
    <http://nagiosplug.sourceforge.net/developer-guidelines.html>,
    Nagios::Plugin::Functions, Nagios::Plugin, RRD::Simple

VERSION
    $Id: RRD.pm 950 2007-02-11 20:40:31Z nicolaw $

AUTHOR
    Nicola Worthington <nicolaw@cpan.org>

    <http://perlgirl.org.uk>

    If you like this software, why not show your appreciation by sending the
    author something nice from her Amazon wishlist? (
    http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

COPYRIGHT
    Copyright 2007 Nicola Worthington.

    This software is licensed under The Apache Software License, Version
    2.0.

    <http://www.apache.org/licenses/LICENSE-2.0>



Changes
Version 0.01 - 12th February 2007
    + Initial developer pre-release to save the namespace
      and to discuss API prior to full release.



Build.PL
# vim:ts=4:sw=4:tw=78
# $Id$

use strict;
use Module::Build;
use vars qw($build);

$build = Module::Build->new(
		module_name        => 'Nagios::Plugin::RRD',
		license            => 'open_source',
		create_makefile_pl => 'passthrough',
		create_readme      => 1,
		create_packlist    => 1,
		sign               => 0,

		requires => {
			'Carp'         => 0, # 1.04
			'Scalar::Util' => 0,
			'RRDs'         => 0,
			'Nagios::Plugin::Functions' => '0.15',
		},

		build_requires => {
			'Test'         => 0, # 1.25
			'Test::More'   => 0, # 0.60
		},

		recommends => {
			'Test::Pod'           => 1.20,
			'Test::Pod::Coverage' => 1.06,
			'Test::Deep'          => 0.093,
		},
	);

$build->create_build_script;


# Send perl and module version information home if we've been given
# permission to do so by a human being - default to not send for automated
# testing environments, of if the user does not respond within 20 seconds.

my $url = $ENV{AUTOMATED_TESTING} ? undef : may_send_version_information();
if ($url) {
	my @resp = ();
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 10;
		my $ua = LWP::UserAgent->new(
				agent => 'Build.PL $Revision$',
				timeout => 9,
				max_size => 500,
			);
		$ua->env_proxy;
		my $response = $ua->get($url);
		if ($response->is_success()) {
			for (split(/\s*\n+\s*/, $response->content())) {
				push @resp, $_ if $_;
			}
		}
		alarm 0;
	};
	print substr($resp[0],0,79) || "Thank you for sending this information.";
	print "\n\n";
}

sub may_send_version_information {
	eval {
		require Config;
		require LWP::UserAgent;
	};
	return undef if $@;

	my $str = sprintf('%s?%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s',
			'http://perlgirl.org.uk/lib/usage.cgi',
			'name',     $build->dist_name(),
			'version',  $build->dist_version(),
			'osname',   $Config::Config{osname},
			'archname', $Config::Config{archname},
			'osver',    $^O,
			'perlver',  $]
		);

	print "\nThank you for downloading ".$build->dist_name()."\n\n";
	print "I would like to find out how many people are using this software,\n";
	print "and on what operating systems and Perl versions. If you have an\n";
	print "internet connection, may I transmit the following information:\n\n";
	print "$str\n\n";

	my $send = 0;
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 20;
		$send = $build->y_n('Send this anonymous information?','n');
		alarm 0;
	};

	return $send ? $str : undef;
}

1;



lib/Nagios/Plugin/RRD.pm
############################################################
#
#   $Id$
#   Nagios::Plugin::RRD - Create RRD threshold Nagios plugins
#
#   Copyright 2007 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Nagios::Plugin::RRD;
# vim:ts=4:sw=4:tw=78

use 5.6.1;
use strict;
use warnings;
use RRDs;
use Nagios::Plugin::Functions qw(nagios_exit nagios_die check_messages);
use Scalar::Util qw(refaddr);
use Carp qw(croak cluck carp confess);
use vars qw($VERSION $DEBUG);

$VERSION = '0.01' || sprintf('%d', q$Revision$ =~ /(\d+)/g);
$DEBUG ||= $ENV{DEBUG} ? 1 : 0;

my $objstore = {};


#
# Public methods
#

sub new {
	ref(my $class = shift) && croak 'Class name required';
	croak 'Odd number of elements passed when even was expected' if @_ % 2;

	# Conjure up an invisible object 
	my $self = bless \(my $dummy), $class;
	$objstore->{refaddr($self)} = {@_};
	my $stor = $objstore->{refaddr($self)};

	# Define what parameters are valid for this constructor
	$stor->{validkeys} = [qw(rrd die_on_unknown)];
	my $validkeys = join('|',@{$stor->{validkeys}});

	# Only accept sensible known parameters from punters
	my @invalidkeys = grep(!/^$validkeys$/,grep($_ ne 'validkeys',keys %{$stor}));
	delete $stor->{$_} for @invalidkeys;
	cluck('Unrecognised parameters passed: '.join(', ',@invalidkeys))
		if @invalidkeys && $^W;

	# Set default values
	$stor->{die_on_unknown} ||= 0;

	# Create empty result sets
	$stor->{unknown}  = [];
	$stor->{critical} = [];
	$stor->{warning}  = [];
	$stor->{ok}       = [];

	# If the RRD file does not exist
	$stor->{rrd} ||= '';
	unless (-e $stor->{rrd}) {
		my $msg = "RRD file '$stor->{rrd}' does not exist";
		push @{$stor->{unknown}}, $msg;
		nagios_die($msg) if $stor->{die_on_unknwon};
	}

	DUMP('$self',$self);
	DUMP('$stor',$stor);
	return $self;
}


sub add_rule {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	my $stor = $objstore->{refaddr($self)};

	# Only accept sensible known parameters from punters
	my @invalidkeys = grep(!/^(ds|rule|result)$/,
		grep($_ ne 'validkeys',keys %{$stor}));
	delete $stor->{$_} for @invalidkeys;
	cluck('Unrecognised parameters passed: '.join(', ',@invalidkeys))
		if @invalidkeys && $^W;


}


sub sources {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	my $stor = $objstore->{refaddr($self)};

	my @ds;
	if (@{$stor->{sources}}) {
		@ds = @{$stor->{sources}};

	} else {
		my $info = RRDs::info($stor->{rrd});
		my $error = RRDs::error();
		croak($error) if $error;

		foreach (keys %{$info}) {
			if (/^ds\[(.+)?\]\.type$/) {
				push @ds, $1;
			}
		}
	}

	return wantarray ? @ds : \@ds;
}


sub last {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	my $stor = $objstore->{refaddr($self)};

	my $last = RRDs::last($stor->{rrd});
	my $error = RRDs::error();
	croak($error) if $error;
	return $last;
}


sub result {
	my $self = shift;
	croak 'Not called as a method by parent object'
		unless ref $self && UNIVERSAL::isa($self, __PACKAGE__);
	my $stor = $objstore->{refaddr($self)};

	return check_messages(
			'critical' => $stor->{critical},
			'warning'  => $stor->{warning},
			'ok'       => $stor->{ok},
			'join'     => ' ',
		);
}





#
# Private methods
#

sub _read_rrd {

}





sub DESTROY {
	my $self = shift;
	delete $objstore->{refaddr($self)};
}


sub TRACE {
	return unless $DEBUG;
	carp(shift());
}


sub DUMP {
	return unless $DEBUG;
	eval {
		require Data::Dumper;
		no warnings 'once';
		local $Data::Dumper::Indent = 2;
		local $Data::Dumper::Terse = 1;
		carp(shift().': '.Data::Dumper::Dumper(shift()));
	}
}


1;



=pod

=head1 NAME

Nagios::Plugin::RRD - Create RRD threshold Nagios plugins

=head1 SYNOPSIS

 use strict;
 use Nagios::Plugin::RRD qw();
 
 my $rrd = Nagios::Plugin::RRD->new(
         rrd => "/var/spool/rrd/host1.acme.com/mailq.rrd"
     );
 
 $rrd->add_rule(
         ds => "",
         threshold => "",
         result => ""
     );
 
 my ($rtn,$msg) = $rrd->result;

=head1 DESCRIPTION

=head1 METHODS

=head2 new

 my $rrd = Nagios::Plugin::RRD->new(
         rrd => "/path/filename.rrd",
         die_on_unknown => 1,
     );

=head2 last

 my $last_updated = $rrd->last;

=head2 sources

 my @sources = $rrd->sources;

=head2 add_rule

 my @sources = $rrd->sources;
 for my $ds (@sources) {
     $rrd->add_rule(
             ds => $ds,
             rule => "",
             result => "CRITICAL",
         );
 }

=head2 result

 my ($rtn,$msg) = $rrd->result;
 print "$msg\n";
 exit $rtn;

=head1 SEE ALSO

L<http://nagiosplug.sourceforge.net/developer-guidelines.html>,
L<Nagios::Plugin::Functions>, L<Nagios::Plugin>, L<RRD::Simple>

=head1 VERSION

$Id$

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

If you like this software, why not show your appreciation by sending the
author something nice from her
L<Amazon wishlist|http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority>? 
( http://www.amazon.co.uk/gp/registry/1VZXC59ESWYK0?sort=priority )

=head1 COPYRIGHT

Copyright 2007 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut


__END__


